From e74e0bdc995ecf63b903df162342aa6bb869c672 Mon Sep 17 00:00:00 2001
From: Shikha Panwar <shikhapanwar@google.com>
Date: Wed, 27 Mar 2024 20:23:31 +0000
Subject: [PATCH 1/2] Claim ownership of VM instance during VM transfer

Apps can transfer a VM instance to another. Claim the ownership of the
VM (at the receiver end) using `claimVmInstance`.

Test: tbd Check Logs
Bug: 294177871
Change-Id: I870a3ff48cb39bdc8be9cfbb7f55aa1148596c44
---
 .../android/system/virtualmachine/VirtualMachine.java | 11 +++++++++++
 virtualizationmanager/src/aidl.rs                     |  7 +++++++
 .../virtualizationservice/IVirtualizationService.aidl |  8 ++++++++
 3 files changed, 26 insertions(+)

diff --git a/java/framework/src/android/system/virtualmachine/VirtualMachine.java b/java/framework/src/android/system/virtualmachine/VirtualMachine.java
index dd2719cff..d166a4d58 100644
--- a/java/framework/src/android/system/virtualmachine/VirtualMachine.java
+++ b/java/framework/src/android/system/virtualmachine/VirtualMachine.java
@@ -433,6 +433,15 @@ public class VirtualMachine implements AutoCloseable {
                 config.serialize(vm.mConfigFilePath);
                 if (vm.mInstanceIdPath != null) {
                     vm.importInstanceIdFrom(vmDescriptor.getInstanceIdFd());
+                    // Notify global VS about the ownership of this instance-id for housekeeping
+                    // purposes.
+                    IVirtualizationService service = vm.mVirtualizationService.getBinder();
+                    try {
+                        byte[] instanceId = Files.readAllBytes(vm.mInstanceIdPath.toPath());
+                        service.claimVmInstance(instanceId);
+                    } catch (IOException e) {
+                        throw new VirtualMachineException("failed to read instance_id", e);
+                    }
                 }
 
                 try {
@@ -461,6 +470,8 @@ public class VirtualMachine implements AutoCloseable {
                 e.addSuppressed(innerException);
             }
             throw e;
+        } catch (RemoteException e) {
+            throw e.rethrowAsRuntimeException();
         }
     }
 
diff --git a/virtualizationmanager/src/aidl.rs b/virtualizationmanager/src/aidl.rs
index b8c63153e..f7128615a 100644
--- a/virtualizationmanager/src/aidl.rs
+++ b/virtualizationmanager/src/aidl.rs
@@ -217,6 +217,7 @@ impl IVirtualizationService for VirtualizationService {
         console_in_fd: Option<&ParcelFileDescriptor>,
         log_fd: Option<&ParcelFileDescriptor>,
     ) -> binder::Result<Strong<dyn IVirtualMachine>> {
+        check_manage_access()?;
         let mut is_protected = false;
         let ret = self.create_vm_internal(
             config,
@@ -231,6 +232,7 @@ impl IVirtualizationService for VirtualizationService {
 
     /// Allocate a new instance_id to the VM
     fn allocateInstanceId(&self) -> binder::Result<[u8; 64]> {
+        check_manage_access()?;
         GLOBAL_SERVICE.allocateInstanceId()
     }
 
@@ -326,6 +328,11 @@ impl IVirtualizationService for VirtualizationService {
         check_manage_access()?;
         Ok(is_secretkeeper_supported())
     }
+
+    fn claimVmInstance(&self, instance_id: &[u8; 64]) -> binder::Result<()> {
+        check_manage_access()?;
+        GLOBAL_SERVICE.claimVmInstance(instance_id)
+    }
 }
 
 impl VirtualizationService {
diff --git a/virtualizationservice/aidl/android/system/virtualizationservice/IVirtualizationService.aidl b/virtualizationservice/aidl/android/system/virtualizationservice/IVirtualizationService.aidl
index 462932c28..ab821a949 100644
--- a/virtualizationservice/aidl/android/system/virtualizationservice/IVirtualizationService.aidl
+++ b/virtualizationservice/aidl/android/system/virtualizationservice/IVirtualizationService.aidl
@@ -98,4 +98,12 @@ interface IVirtualizationService {
      * For more info see packages/modules/Virtualization/docs/updatable_vm.md
      */
     boolean isUpdatableVmSupported();
+
+    /**
+     * Notification that ownership of a VM has been claimed by the caller.  Note that no permission
+     * checks (with respect to the previous owner) are performed.
+     *
+     * @param instanceId The ID for the VM.
+     */
+    void claimVmInstance(in byte[64] instanceId);
 }
-- 
2.44.0.478.gd926399ef9-goog

